syscalls.c

```diff
@@ -316,23 +317,31 @@ void syscalls_meminfo(void *ustack)
 int syscalls_syspageprog(void *ustack)
 {
 	int i;
-	size_t sz;
+	size_t sz, namesz;
 	syspageprog_t *prog;
+	const syspage_prog_t *progSys;
 
 	GETFROMSTACK(ustack, syspageprog_t *, prog, 0);
 	GETFROMSTACK(ustack, int, i, 1);
 
+	sz = syspage_progSize();
 	if (i < 0)
-		return syspage->progssz;
+		return sz;
+
+	if (i >= sz)
+		return -EINVAL;
 
-	if (i >= syspage->progssz)
+	if ((progSys = syspage_progIdResolve(i)) == NULL)
 		return -EINVAL;
 
-	prog->addr = syspage->progs[i].start;
-	prog->size = syspage->progs[i].end - syspage->progs[i].start;
-	sz = min(sizeof(prog->name), sizeof(syspage->progs[i].cmdline));
-	hal_memcpy(prog->name, syspage->progs[i].cmdline, sz);
-	prog->name[sz - 1] = '\0';
+	prog->addr = progSys->start;
+	prog->size = progSys->end - progSys->start;
+
+	/* TODO: change syspageprog_t to allocate data for name dynamically */
+	namesz = hal_strlen(progSys->argv);
+	sz = min(sizeof(prog->name), namesz);
```

Comment:
```suggestion
	sz = min(sizeof(prog->name) - 1, namesz);
```
if `sz == sizeof(prog->name)` then `prog->name[sz] ` is out off buffer.

Reply:
Yeah, thanks for catching this one.